### ISSUE TITLE
Migrate to Manifest V3

### ISSUE THREAD

**Issue by b0o** [brookhong/Surfingkeys#1821]
> To use chrome.tabGroups APIs, the extension needs to [migrate to Manifest V3](https://developer.chrome.com/docs/extensions/mv3/mv3-migration-checklist/).
**But using Manifest V3 will break the feature that Surfingkeys allows to execute user scripts, which is important for customization.**

&mdash; [gkzhb](https://github.com/gkzhb) on https://github.com/brookhong/Surfingkeys/issues/1445#issuecomment-939479553

Bolding mine - need to investigate this claim that V3 will break the ability to execute user scripts.

---

> [Support for Manifest V2 will be removed from Chrome in January 2023](https://developer.chrome.com/docs/extensions/mv3/mv2-sunset/), which is 4 months away, so Surfingkeys needs to decide how to handle that quite urgently anyway - I don't think "wait and see" is a viable option any more. Should there be a separate issue for tracking MV3 support?

&mdash; [aspiers](https://github.com/aspiers) on https://github.com/brookhong/Surfingkeys/issues/1445#issuecomment-1234593488

--- 

### Update 09/28/22

MV2 deprecation in Chrome Stable has been postponed until June 2023. 

> - Starting in January in Chrome 112, Chrome may run experiments to turn off support for Manifest V2 extensions in Canary, Dev, and Beta channels.
> - Starting in June in Chrome 115, Chrome may run experiments to turn off support for Manifest V2 extensions in all channels, including stable channel.
> - In June 2023, the Chrome Web Store will no longer allow Manifest V2 items to be published with [visibility](https://developer.chrome.com/docs/webstore/cws-dashboard-distribution/#setting-the-visibility) set to Public. All existing Manifest V2 items with visibility set to Public at that time will have their visibility changed to Unlisted.
> - In January 2024, following the expiration of the Manifest V2 enterprise policy, the Chrome Web Store will remove all remaining Manifest V2 items from the store.

See: [More details on the transition to Manifest V3](https://developer.chrome.com/blog/more-mv2-transition/)

**Comment by b0o** [comment]
Relevant discussion: https://github.com/Tampermonkey/tampermonkey/issues/644

**Comment by b0o** [comment]
[Google postpones MV2 shutoff in Chrome stable to June 2023](https://developer.chrome.com/blog/more-mv2-transition/)

**Comment by brookhong** [comment]
Some API changes are needed for manifest upgrade, see [Migrate chrome extension from manifest v2 to v3: How to update these APIs? - Stack Overflow](https://stackoverflow.com/questions/71534432/migrate-chrome-extension-from-manifest-v2-to-v3-how-to-update-these-apis).

**Comment by brookhong** [comment]
I had looked into this, now the blocker is that background service worker just cannot start, it remains `Inactive`. To reload the extension does not help either. The worse thing is that there is no error or warning message. It just failed silently.

<img width="405" alt="image" src="https://user-images.githubusercontent.com/288207/202857253-494a5e93-5454-4971-aac9-d7c293b52e2f.png">

Looks like [the bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1271154) just got fixed 4 days ago, will try that again after the fix released.

**Comment by brookhong** [comment]
Pending on https://github.com/w3c/webextensions/issues/279#issuecomment-1329586697

**Comment by EvanCarroll** [comment]
This deadline is quickly approaching. The new dates are,

> * In June 2023, the Chrome Web Store will no longer allow Manifest V2 items to be published with [visibility](https://developer.chrome.com/docs/webstore/cws-dashboard-distribution/#setting-the-visibility) set to Public. All existing Manifest V2 items with visibility set to Public at that time will have their visibility changed to Unlisted.
> * In January 2024, following the expiration of the Manifest V2 enterprise policy, the Chrome Web Store will remove all remaining Manifest V2 items from the store.

**Comment by asknet** [comment]
@brookhong @b0o Appreciate it if the team could share insights on the current state. Thanks!

**Comment by b0o** [comment]
> This deadline is quickly approaching. The new dates are,
> 
> > * In June 2023, the Chrome Web Store will no longer allow Manifest V2 items to be published with [visibility](https://developer.chrome.com/docs/webstore/cws-dashboard-distribution/#setting-the-visibility) set to Public. All existing Manifest V2 items with visibility set to Public at that time will have their visibility changed to Unlisted.
> > * In January 2024, following the expiration of the Manifest V2 enterprise policy, the Chrome Web Store will remove all remaining Manifest V2 items from the store.

I believe the Chrome team pushed back these dates indefinitely. That’s not to say depreciation wont happen, but there are no official dates as far as I can tell. 

> We're still working on the timeline of the MV2 phase-out plan, so keep an eye out for it in the coming months. We will provide at least 6 months between a timeline announcement and any experiments deprecating MV2 features.
> Before announcing a new timeline, we are addressing prioritized platform gaps and closing critical bugs. You can track these on the [Known Issues](https://developer.chrome.com/docs/extensions/migrating/known-issues/) page.

https://developer.chrome.com/docs/extensions/migrating/mv2-sunset/

**Comment by asknet** [comment]
@b0o Thank you for the update

**Comment by EvanCarroll** [comment]
I don't read that the same way, so be cautious here. Those **are the updated deadlines** as I understand it, and as reported by third parties. From the article titled, _"Google delays execution of doomed Chrome extensions"_

> New MV2 extensions could no longer be added to the Chrome Web Store in June 2022, and that remains unchanged under the new roadmap; MV2 extensions already available the Chrome Web Store can still be downloaded and can still receive updates.
>
> As of June 2023, MV2 extensions will no longer be visible in the store (so they can't be newly installed, but can still be updated for existing users).
>
> **Come January 2024, nothing will be left to chance: the Chrome Web Store will stop accepting updates to MV2 extensions, all MV2 extensions will be removed from the store, and the MV2 usage in enterprises will end.**

https://www.theregister.com/2022/09/30/google_manifestv3_chrome_extensions/

**Comment by b0o** [comment]
@EvanCarroll That article is dated 30 Sep 2022, it's about the first time Google delayed MV3. They later delayed it again, here is an article dated 01 Apr 2023:

> "We're still working on the timeline of the MV2 phase-out plan, so keep an eye out for it in the coming months. We will provide sufficient migration time for developers – at least six months of heads-up – before beginning any experiments to turn off MV2 in the browser next year."

https://www.theregister.com/2023/04/01/google_mv2_chrome_extensions/

That article is based on this post on Google Groups:

> Oliver Dunk
Mar 29, 2023, 10:44:50 AM
to Chromium Extensions, Simeon Vincent
Hey everyone,
> 
> As promised, we wanted to share an update on how things are going. Over the last few months, we’ve made progress on a number of improvements based on developer community feedback, such as allowing extension events to [extend a service worker’s lifetime](https://developer.chrome.com/blog/longer-esw-lifetimes/), [increasing in-memory session storage](https://bugs.chromium.org/p/chromium/issues/detail?id=1385167) from 1 MB to 10 MB [after cross-browser discussions](https://github.com/w3c/webextensions/issues/350), and adding more reasons for [offscreen documents](https://developer.chrome.com/docs/extensions/reference/offscreen). We're [continuing to improve](https://developer.chrome.com/docs/extensions/mv3/known-issues/) the platform by making service workers easier to use (including more flexible lifetimes with messaging and other APIs), introducing more service worker APIs in general, and working with other browser vendors to find ways we can align our extension platforms.  
> 
> We plan to continue reviewing feedback, making changes and improving documentation to ensure the transition from Manifest V2 to Manifest V3 is smooth and successful. We're still working on the timeline of the MV2 phase-out plan, so keep an eye out for it in the coming months. We will provide sufficient migration time for developers - at least 6 months of heads-up - before beginning any experiments to turn off MV2 in the browser next year.
> 
> We remain committed to the rollout of MV3 to improve security, privacy, and performance for our users around the world. We also want to ensure that we phase out MV2 in a timely manner, and we’re continuing to listen to feedback from the developer community to help inform our improvements and timelines.
> 
> Thanks for your patience while we continue to work on moving the extensions ecosystem forward in a way that supports the needs of users and developers.
> 
> Chrome Extensions Team

https://groups.google.com/a/chromium.org/g/chromium-extensions/c/zQ77HkGmK9E/m/HjaaCIG-BQAJ (If you're interested in this I recommend subscribing to that thread)

As far as I know, there haven't been any additional updates, so I think it should be at least 6+ months away. That's also what the official sunset page says:

> We're still working on the timeline of the MV2 phase-out plan, so keep an eye out for it in the coming months. We will provide at least 6 months between a timeline announcement and any experiments deprecating MV2 features.

https://developer.chrome.com/docs/extensions/migrating/mv2-sunset/

**Comment by EvanCarroll** [comment]
Thanks, that Mar 29th post is certainly the newest so I'm convinced. Talk about a butchered roll out. They really googled that one up.

**Comment by brookhong** [comment]
Yes, google must have delayed this again.

As for this extension, I had created a branch https://github.com/brookhong/Surfingkeys/tree/mv3 last December, it roughly works(I haven't verified every function), so MV3 should not be a big problem as long as a string parameter `code` could be supported by the `scripting.executeScript()` to inject user-script into context of the extension itself, otherwise we will lose the function to use user snippets to customize settings -- creating users' own bindings, which would definitely be a big loss.

I check https://github.com/w3c/webextensions/issues/279#issuecomment-1329586697 occasionally, there is no update this year except more and more extensions coming and requesting for the same support.

**Comment by b0o** [comment]
Another update from Google: https://developer.chrome.com/blog/resuming-the-transition-to-mv3/

**Comment by asknet** [comment]
@newRoland I don't know the full architecture. With this workaround, what are we losing?

**Comment by EvanCarroll** [comment]
Thanks @b0o for dropping the link. This is such a botched migration. Google needs help, badly. Here is the tldr of this release, Google will be **disabling v2 in Dev, Canary, and Beta in June**, and they'll adapt that to stifle backlash until they're comfortable and then at some undetermined point they intend to graduate that functionality to production, in stages.

**Comment by newRoland** [comment]
@asknet Surfingkeys allows users to execute user scripts. No loss if you migrate and use the wrapped userScript approach to maintain this feature. The wrapped userScript approach needs to be implemented with care as it's performance isn't great. The user script should be scoped to a particular website because you're pretty much pre-loading. If not scoped to a particular URL, you're pretty much loading it on all websites, even if it's irrelevant and will never be called.

**Comment by newRoland** [comment]
I have a great update! It looks like the proposal is moving forward. 

https://github.com/w3c/webextensions/pull/540

**Comment by macintacos** [comment]
Heya folks, sorry, it's not clear to me what's supposed to be happening next. The proposal linked above seems to have been merged a few months ago, what are the next steps here? Really would be sad if I was suddenly unable to use Surfingkeys 😅

**Comment by brookhong** [comment]
I have been working on this for some while, here are some updates

1. The biggest challenge now is to make setting snippet work as in MV2. The setting snippet is just a piece of user script that needs to be executed in the world/context of extension, MV3 does not allow that. Though [chrome.userScripts.register](https://developer.chrome.com/docs/extensions/reference/api/userScripts#type-RegisteredUserScript) can help to inject any  user script, but the script can only be injected into a `USER_SCRIPT` world or `MAIN` world, both of which don't have the necessary functions/objects that need to be used in setting snippet to enable users to create their own mappings/customize settings. Thus we need implement a mechanism to relay those function calls in setting snippet to the extension world, fortunately it is possible, but it still take some efforts to cover all functions/objects that need to be available in setting snippet.
2. Above solution may not work for non-chromium based browsers, so I will probably use [mv3](https://github.com/brookhong/Surfingkeys/tree/mv3) for Chrome release, and merge it back to mainline after MV3 becomes fully ready for other browsers.

**Comment by brookhong** [comment]
1.17.0 is released to migrate to MV 3, please expect that there might be something broken for you though I tested it completely in my Chrome.

**Comment by petobens** [comment]
Hi @brookhong . In 1.17 I cannot seem to load Load Settings by pointing to a file. Is this expected?
![image](https://github.com/user-attachments/assets/623c3c78-9c43-4864-ba5b-9eed39693386)

**Comment by brookhong** [comment]
It works for me after `Allow access to file URLs` is turned on.
![image](https://github.com/user-attachments/assets/12681ba6-4d9e-452f-bb5d-d6c724071f64)

**Comment by laphylee** [comment]
@brookhong I met the same problems: can't set the customized file path even after I turn on "Allow access to file URLs".

**Comment by fecet** [comment]
I can read config file from local or remote, but it keep report
```
[SurfingKeys] Error found in settings: TypeError: Cannot read properties of undefined (reading 'local')`\
```

**Comment by clarkwang** [comment]
> 1.17.0 is released(just for Chrome) to migrate to MV 3, please expect that there might be something broken for you though I tested it completely in my Chrome.

I hit this error with v1.17:

~~~
[SurfingKeys] Error found in settings: TypeError: api.Hints.setCharacters is not a function
~~~

Seems like `api.Hints.setCharacters` is now a var and this works for me:

~~~
api.Hints.setCharacters = 'abcdefhjklmnoprstuvwxyz';
~~~

**Comment by brookhong** [comment]
@fecet please check your config file and find where `local` is being referred.

@clarkwang yes, the api `setCharacters` for `Hint` is missed to be exported. Though your fix suppresses the error, but it does not actually do the job. This will be fixed soon. Thanks for reporting.

Looks like you guys can load your config file, while @petobens and @laphylee can't.

@petobens Could you try `file:///home/<login>/.surfingkeysrc`?

**Comment by petobens** [comment]
> @petobens Could you try file:///home/<login>/.surfingkeysrc

No success. For the record I'm tried this both in `Brave Browser 128.1.69.168` and `Microsoft Edge 130.0.2808.0 dev`

**Comment by fecet** [comment]
> @fecet please check your config file and find where `local` is being referred.

I don't find any `local` in my config except `chrome.storage.local`, does it matter? And I find
```
Uncaught (in promise) Error: Could not establish connection. Receiving end does not exist.
```
at [chrome://extensions/?errors=gfbliohnnapiefjpjlpjnehglfpaknnc](chrome://extensions/?errors=gfbliohnnapiefjpjlpjnehglfpaknnc)

**Comment by jackmac92** [comment]
Getting this error after the update `[SurfingKeys] Error found in settings: TypeError: api.vmapkey is not a function`

**Comment by brookhong** [comment]
> > @fecet please check your config file and find where `local` is being referred.
> 
> I don't find any `local` in my config except `chrome.storage.local`, does it matter? And I find
> 
> ```
> Uncaught (in promise) Error: Could not establish connection. Receiving end does not exist.
> ```
> 
> at chrome://extensions/?errors=gfbliohnnapiefjpjlpjnehglfpaknnc

Yeah, that's the problem, `chrome.storage` is not allowed in user script, which is a restriction from Chrome.

**Comment by brookhong** [comment]
> Getting this error after the update `[SurfingKeys] Error found in settings: TypeError: api.vmapkey is not a function`
> 
> edit: but the rest of my config still works which is awesome!! thanks for the work on this @brookhong

Thanks, `vmapkey` is missed from exporting, will be fixed soon.

**Comment by laphylee** [comment]
@brookhong I tried both "d:\OneDriveShare\Setting\Surfingkeys.js" and "file:///d:/OneDriveShare/Setting/Surfingkeys.js" but not successful. 
I got this error on console log: 
![image](https://github.com/user-attachments/assets/4ea85977-9fff-4ce3-b242-00ffa814ccb2)

**Comment by brookhong** [comment]
For anyone who is unable to load his/her config file, please try @matko031's solution https://github.com/brookhong/Surfingkeys/issues/2149#issuecomment-2354694822, just to remove the extension then re-add it.

**Comment by deadcyclo** [comment]
> For anyone who is unable to load his/her config file, please try @matko031's solution [#2149 (comment)](https://github.com/brookhong/Surfingkeys/issues/2149#issuecomment-2354694822), just to remove the extension then re-add it.

I tried this workaround both in chrome and in vivaldi and it didn't work in either. And no error messages in console.

**Comment by paulexyz** [comment]
Same here, just tried to install Surfingkeys in a new Chrome profile (Version 1.17.0)
And I can't access a file with my settings: `Cannot read properties of null (reading 'getValue')`

Chrome version:
`128.0.6613.84 (Official Build) (64-bit) `

**Comment by lost22git** [comment]
Same error info:

![Screenshot 2024-09-18 004257](https://github.com/user-attachments/assets/79f0bd9c-40c5-44f4-b92f-02f8159c2638)

and cannot add proxy pair at the same time.

**Comment by bojanjovicic** [comment]
> Same error info when save configuration:
> 
> ![Screenshot 2024-09-18 004257](https://private-user-images.githubusercontent.com/65008815/368256322-79f0bd9c-40c5-44f4-b92f-02f8159c2638.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY1OTk1MzgsIm5iZiI6MTcyNjU5OTIzOCwicGF0aCI6Ii82NTAwODgxNS8zNjgyNTYzMjItNzlmMGJkOWMtNDBjNS00NGY0LWI5MmYtMDJmODE1OWMyNjM4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE3VDE4NTM1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY1ODFiOTgxNTg0ZWNiMjIyNTBjZTdmMTljZGI0MzE5YzJjMTQyNzEzMDc1OTM1YjUzNjlkZTliNDJhMDE5M2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DdF_aAlQZzpC4gXeXoG-vyU3tJXIi40EOB6wIeTol64)
> 
> and also cannot add proxy pair.
> 
> Extension seems no error info hints:
> 
> ![Screenshot 2024-09-18 004257](https://private-user-images.githubusercontent.com/65008815/368257435-3468ae36-d0a1-4c65-9094-97939dd30c41.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY1OTk1MzgsIm5iZiI6MTcyNjU5OTIzOCwicGF0aCI6Ii82NTAwODgxNS8zNjgyNTc0MzUtMzQ2OGFlMzYtZDBhMS00YzY1LTkwOTQtOTc5MzlkZDMwYzQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE3VDE4NTM1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBjY2ZlYjdmYWI2ZTRlNjdmMGQ3YTgyZjA2MjIyM2YxZDYxNzk1ZWUyNjRkNDRlYzUyMzBhODU3NTgwNjcwODAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TT9wySoYnmCPJfFi6uHA5emNXHkCIztmT-it0FLRskM)

Same here: 
![image](https://github.com/user-attachments/assets/4fff6db8-1b57-40ec-aa29-9cab1dc68501)

**Comment by brookhong** [comment]
Closing this now, and please report your broken use cases here with https://github.com/brookhong/Surfingkeys/issues/2159.
